### PR TITLE
Fix DeleteOptions on Public Ip Address (issue#38806)

### PR DESCRIPTION
**Pull request by v-hongli1**
relate to [issue#38806](https://github.com/Azure/azure-sdk-for-java/issues/38806)

# Description

Please add an informative description that covers that changes made by the pull request and link all relevant issues.

If an SDK is being regenerated based on a new swagger spec, a link to the pull request containing these swagger spec changes has been included above.

# All SDK Contribution checklist:
- [ ] **The pull request does not introduce [breaking changes]**
- [ ] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [ ] **I have read the [contribution guidelines](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md).**

## [General Guidelines and Best Practices](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md#developer-guide)
- [ ] Title of the pull request is clear and informative.
- [ ] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### [Testing Guidelines](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md#building-and-unit-testing)
- [ ] Pull request includes test coverage for the included changes.

### CODE REVIEW

File: sdk/resourcemanager/azure-resourcemanager-network/src/main/java/com/azure/resourcemanager/network/implementation/NetworkInterfaceImpl.java [hunk 1]
```diff
@@ -416,6 +418,22 @@ public NetworkSecurityGroup getNetworkSecurityGroup() {
...
 
+    @Override
+    public NicIpConfigurationImpl specifiedIPConfiguration(String ipConfigName) {
+        NicIpConfigurationImpl specifiedIPConfig = null;
+        for (NicIpConfiguration ipConfiguration : this.nicIPConfigurations.values()) {
+            if (ipConfigName.equals(ipConfiguration.name())) {
+                specifiedIPConfig = (NicIpConfigurationImpl) ipConfiguration;
+                break;
+            }
+        }
```
Comment (weidongxu-microsoft) [inline sdk/resourcemanager/azure-resourcemanager-network/src/main/java/com/azure/resourcemanager/network/implementation/NetworkInterfaceImpl.java:429]: This part is just `ipConfigurations().get(ipConfigName)`?

File: sdk/resourcemanager/azure-resourcemanager-network/src/main/java/com/azure/resourcemanager/network/implementation/NetworkInterfaceImpl.java [hunk 2]
```diff
@@ -416,6 +418,22 @@ public NetworkSecurityGroup getNetworkSecurityGroup() {
...
+        if (specifiedIPConfig == null) {
+            specifiedIPConfig = prepareNewNicIPConfiguration(ipConfigName);
+            withIPConfiguration(specifiedIPConfig);
+        }
```
Comment (weidongxu-microsoft) [inline sdk/resourcemanager/azure-resourcemanager-network/src/main/java/com/azure/resourcemanager/network/implementation/NetworkInterfaceImpl.java:433]: No, we don't want to modify anything in a getter API.

**Review by weidongxu-microsoft — COMMENTED** [review]
There is too many issue in the PR.

Please book a meeting next week to discuss.

File: sdk/resourcemanager/azure-resourcemanager-network/src/main/java/com/azure/resourcemanager/network/models/NetworkInterface.java [hunk 3]
```diff
@@ -29,6 +29,14 @@ public interface NetworkInterface
     /** @return the primary IP configuration of this network interface */
     NicIpConfiguration primaryIPConfiguration();
 
+    /**
+     * Gets the specified ip configuration
+     *
+     * @param ipConfigName the name of th specified ip configuration
+     * @return the specified IP configuration of this network interface
+     */
+    NicIpConfiguration specifiedIPConfiguration(String ipConfigName);
```
Comment (XiaofeiCao) [inline sdk/resourcemanager/azure-resourcemanager-network/src/main/java/com/azure/resourcemanager/network/models/NetworkInterface.java:38]: This is equivalent to `ipConfigurations().get(ipoConfigName)`?

File: sdk/resourcemanager/azure-resourcemanager-network/src/main/java/com/azure/resourcemanager/network/implementation/NetworkInterfaceImpl.java [hunk 4]
```diff
@@ -563,9 +581,24 @@ protected void beforeCreating() {
...
+        return this;
+    }
+
+    @Override
+    public NetworkInterfaceImpl withSpecifiedNicSpecifiedPipDeleteOption(
+        DeleteOptions deleteOptions, PublicIpAddress publicIpAddress, String ipConfigName) {
+        this.specifiedIPConfiguration(ipConfigName).withExistingPublicIpAddress(publicIpAddress);
+        this.specifiedIPConfiguration(ipConfigName).withPrivateIpVersion(publicIpAddress.version());
+        this.publicIPAddressDeleteOption = deleteOptions;
+        return this;
```
Comment (XiaofeiCao) [inline sdk/resourcemanager/azure-resourcemanager-network/src/main/java/com/azure/resourcemanager/network/implementation/NetworkInterfaceImpl.java:602]: The implementation seems to differ from the interface declaration? 

In interface, it's `Update delete option for specified public ip address of specified network interface ip config`, but here it seems to attach an existing PublicIpAddress to the NetworkInterface.

File: sdk/resourcemanager/azure-resourcemanager-compute/src/main/java/com/azure/resourcemanager/compute/models/VirtualMachine.java [hunk 5]
```diff
@@ -2610,6 +2622,33 @@ interface WithDeleteOptions {
...
+
+            /**
+             * Update delete option for the specified public ip address of the specified network interface.
+             *
+             * @param deleteOptions what happens to the public IP address when the VM using it is deleted
+             * @param nicId resource id of the specified network interfac
+             * @param ipConfigIds resource Ids of specified network interface ip configs
+             * @return the next stage of the update
+             */
+            Update withSpecifiedNicSpecifiedPIpDeleteOption(DeleteOptions deleteOptions, String nicId, String... ipConfigIds);
```
Comment (XiaofeiCao) [inline sdk/resourcemanager/azure-resourcemanager-compute/src/main/java/com/azure/resourcemanager/compute/models/VirtualMachine.java:2650]: Not sure whether user really needs these APIs.

IMHO, two are needed:
1. `withPublicIPAddressDeleteOptions(DeleteOptions deleteOptions)` to update all attached PublicIpAddress'es deleteOption
2. `withPublicIPAddressDeleteOption(PublicIpAddress(/String) publicIpAddress(id))`, like what user specifies in the creation flow.

File: sdk/resourcemanager/azure-resourcemanager-compute/src/test/java/com/azure/resourcemanager/compute/VirtualMachineOperationsTests.java [hunk 6]
```diff
@@ -1998,6 +2002,345 @@ private CreatablesInfo prepareCreatableVirtualMachines(
...
+                .withDynamicIP()
+                .create();
+
+        VirtualMachine vm = computeManager.virtualMachines()
+            .define(vmName)
+            .withRegion(region)
+            .withNewResourceGroup(rgName)
+            .withNewPrimaryNetwork("10.0.0.0/28")
+            .withPrimaryPrivateIPAddressDynamic()
+            .withExistingPrimaryPublicIPAddress(publicIPAddress)
```
Comment (XiaofeiCao) [inline sdk/resourcemanager/azure-resourcemanager-compute/src/test/java/com/azure/resourcemanager/compute/VirtualMachineOperationsTests.java:2022]: Should we do `.withExistingPrimaryPublicIPAddress(publicIPAddress, deleteOptions)`?

File: sdk/resourcemanager/azure-resourcemanager-compute/src/test/java/com/azure/resourcemanager/compute/VirtualMachineOperationsTests.java [hunk 7]
```diff
@@ -1998,6 +2002,345 @@ private CreatablesInfo prepareCreatableVirtualMachines(
...
+            .withPopularWindowsImage(KnownWindowsVirtualMachineImage.WINDOWS_SERVER_2019_DATACENTER_GEN2)
+            .withAdminUsername("Foo12")
+            .withAdminPassword(password())
+            .withNewDataDisk(127)
+            .withSize(VirtualMachineSizeTypes.fromString("Standard_D2a_v4"))
+            .withPublicIPAddressDeleteOptions(DeleteOptions.DELETE)
```
Comment (XiaofeiCao) [inline sdk/resourcemanager/azure-resourcemanager-compute/src/test/java/com/azure/resourcemanager/compute/VirtualMachineOperationsTests.java:2028]: Not sure whether we should use this in creation flow. I think it's for update flow.

File: sdk/resourcemanager/azure-resourcemanager-network/src/test/java/com/azure/resourcemanager/network/NetworkInterfaceOperationsTests.java [hunk 8]
```diff
@@ -509,6 +512,116 @@ public void canAssociateNatGateway() {
...
+            .withSubnet(subnetName)
+            .withPrimaryPrivateIPAddressDynamic()
+            .withExistingPrimaryPublicIPAddress(publicIPAddress1)
+            .defineSecondaryIPConfiguration("secondary")
+            .withExistingNetwork(vnet)
+            .withSubnet(subnetName)
+            .withPrivateIpAddressDynamic()
+            .withExistingPublicIpAddress(publicIPAddress2)
+            .attach()
+            .withPublicIPAddressDeleteOptions(DeleteOptions.DELETE)
```
Comment (XiaofeiCao) [inline sdk/resourcemanager/azure-resourcemanager-network/src/test/java/com/azure/resourcemanager/network/NetworkInterfaceOperationsTests.java:558]: Update flow.

File: sdk/resourcemanager/azure-resourcemanager-network/src/test/java/com/azure/resourcemanager/network/NetworkInterfaceOperationsTests.java [hunk 9]
```diff
@@ -509,6 +512,116 @@ public void canAssociateNatGateway() {
...
+            .create();
+
+        NetworkInterface nic = networkManager.networkInterfaces()
+            .define(generateRandomResourceName("nic-", 15))
+            .withRegion(Region.US_EAST)
+            .withNewResourceGroup(rgName)
+            .withExistingPrimaryNetwork(vnet)
+            .withSubnet(subnetName)
+            .withPrimaryPrivateIPAddressDynamic()
+            .withExistingPrimaryPublicIPAddress(publicIPAddress1)
```
Comment (XiaofeiCao) [inline sdk/resourcemanager/azure-resourcemanager-network/src/test/java/com/azure/resourcemanager/network/NetworkInterfaceOperationsTests.java:551]: `withExistingPrimaryPublicIPAddress(publicIPAddress1, deleteOption)`

File: sdk/resourcemanager/azure-resourcemanager-network/src/test/java/com/azure/resourcemanager/network/NetworkInterfaceOperationsTests.java [hunk 10]
```diff
@@ -509,6 +512,116 @@ public void canAssociateNatGateway() {
...
+            .defineSecondaryIPConfiguration("secondary")
+            .withExistingNetwork(vnet)
+            .withSubnet(subnetName)
+            .withPrivateIpAddressDynamic()
+            .withExistingPublicIpAddress(publicIPAddress2)
```
Comment (XiaofeiCao) [inline sdk/resourcemanager/azure-resourcemanager-network/src/test/java/com/azure/resourcemanager/network/NetworkInterfaceOperationsTests.java:556]: `withExistingPublicIpAddress(publicIPAddress2, deleteOption)`?

File: sdk/resourcemanager/azure-resourcemanager-network/src/main/java/com/azure/resourcemanager/network/models/NetworkInterface.java [hunk 11]
```diff
@@ -256,6 +264,17 @@ interface WithAcceleratedNetworking {
...
 
+        /** The stage of the definition allowing to specify delete options for the public ip address. */
+        interface WithPublicIPAddressDeleteOptions {
+            /**
+             * Sets delete options for public ip address.
+             *
+             * @param deleteOptions the delete options for primary network interfaces
+             * @return the next stage of the definition
+             */
+            WithCreate withPublicIPAddressDeleteOptions(DeleteOptions deleteOptions);
```
Comment (XiaofeiCao) [inline sdk/resourcemanager/azure-resourcemanager-network/src/main/java/com/azure/resourcemanager/network/models/NetworkInterface.java:275]: Update flow?

File: sdk/resourcemanager/azure-resourcemanager-network/src/main/java/com/azure/resourcemanager/network/models/NetworkInterface.java [hunk 12]
```diff
@@ -576,6 +596,19 @@ interface WithLoadBalancer {
...
+        interface WithSpecifiedNicSpecifiedPipDeleteOption {
+            /**
+             * Update delete option for specified public ip address of specified network interface ip config.
+             *
+             * @param deleteOptions what happens to the public IP address when the VM using it is deleted
+             * @param publicIpAddress an object of existing public ip address
+             * @param ipConfigName the name of specified network interface ip config
+             * @return the next stage of the update
+             */
+            Update withSpecifiedNicSpecifiedPipDeleteOption(DeleteOptions deleteOptions, PublicIpAddress publicIpAddress, String ipConfigName);
```
Comment (XiaofeiCao) [inline sdk/resourcemanager/azure-resourcemanager-network/src/main/java/com/azure/resourcemanager/network/models/NetworkInterface.java:610]: `withPublicIpAddressDeleteOption(PublicIpAddress pip, DeleteOptions deleteOption)?

**Pull request closed without merging**